docs/source/configuration.rst
=============
Configuration
=============

DXR learns how to index and serve your source trees by means of an ini-formatted
configuration file:

.. include:: example-configuration.rst

When you invoke :program:`dxr index`, it defaults to reading :file:`dxr.config`
in the current directory::

    dxr index

Or you can pass in a config file explicitly::

    dxr index --config /some/place/dxr.config


Sections
========

The configuration file is divided into sections. The ``[DXR]`` section holds
global options; each other section describes a tree to be indexed.

You can use all the fancy interpolation features of Python's
`ConfigParser <http://docs.python.org/library/configparser.html>`__ class to
save repetition.

[DXR] Section
-------------

Here are the options that can live in the ``[DXR]`` section. For options
representing path names, relative paths are relative to the directory
containing the config file.

``disabled_plugins``
    Names of plugins to disable. Default: empty

``enabled_plugins``
    Names of plugins to enable. Default: ``*``

``es_alias``
    A ``format()``-style template for coming up with elasticsearch alias
    names. These live in the same namespace as indices, so don't pave over any
    index name you're already using. The variables ``{format}`` and ``{tree}``
    will be substituted, and their meanings are as in ``es_index``. Default:
    ``dxr_{format}_{tree}``.

``es_index``
    A ``format()``-style template for coming up with elasticsearch index
    names. The variable ``{tree}`` will be replaced with the tree name,
    ``{format}`` will be replaced with the format version, and ``{unique}``
    will be replaced with a unique ID to keep a tree's new index from
    colliding with the old one. The unique ID includes a random number and the
    build hosts's MAC address so errant concurrent builds on different hosts
    at least won't clobber each other. Default: ``dxr_{format}_{tree}_{unique}``

``es_catalog_replicas``
    The number of elasticsearch replicas to make of the :term:`catalog index`.
    This is read often and written only when an indexing run completes, so
    crank it up so there's a replica on every node for best performance. But
    remember that writes will hang if at least half of the attempted copies
    aren't available. Default: ``1``

``es_indexing_timeout``
    The number of seconds DXR should wait for elasticsearch responses during
    indexing. Default: 60

``es_indexing_retries``
    How many other ES nodes to try if a query to one during indexing times out
    or the connection fails. This is an experimental feature. Default: 0

``es_refresh_interval``
    The number of seconds between elasticsearch's consolidation passes during
    indexing. Set to -1 to do no refreshes at all, except directly after an
    indexing run completes. Default: 60

``generated_date``
    The "generated on" date stamped at the bottom of every DXR web page, in
    RFC-822 (also known as RFC 2822) format. Default: the time the indexing run
    started

``log_folder``
    A ``format()``-style template for deciding where to store log files
    written while indexing. The token ``{tree}`` will be replaced with the name
    of the tree being indexed. Default: ``dxr-logs-{tree}`` (in the current
    working directory).

``skip_stages``
    Build/indexing/clean stages to skip, for debugging: ``build``, ``index``,
    ``clean``, or any combination, whitespace-separated Either of ``build`` or
    ``index`` implies ``clean``. Default: none

``temp_folder``
    A ``format()``-style template for deciding where to store temporary files
    used while indexing. The token ``{tree}`` will be replaced with the name
    of each tree you index. Default: ``dxr-temp-{tree}``. It's a good idea to
    keep this out of :file:`/tmp` if it's on a small partition, since it can
    grow to tens of gigabytes on a large codebase.

``workers``
    Number of concurrent processes to use for building and indexing projects.
    Default: the number of CPUs on the system. Set to 0 to use no worker
    processes and do everything in the master process. This is handy for
    debugging.

Web App Options That Need a Restart
```````````````````````````````````

These options are used by the DXR web app (though some are used at index time
as well). They are not frozen into the :term:`catalog index` but rather are
read when the web app starts up. Thus, the web app must be restarted to see
new values of these.

``default_tree``
    The tree to redirect to when you visit the root of the site. Default: the
    first tree in the config file

``es_hosts``
    A whitespace-delimited list of elasticsearch nodes to talk to. Be sure to
    include port numbers. Default: http://127.0.0.1:9200/. Remember that you
    can split whitespace-containing things across lines in an ini file by
    leading with spaces.

``es_catalog_index``
     The name to use for the :term:`catalog index`. You probably don't need to
     change this unless you want multiple otherwise-independent DXR
     deployments, with disjoint Switch Tree menus, sharing the same ES
     cluster. Default: ``dxr_catalog``.

``google_analytics_key``
    Google analytics key. If set, the analytics snippet will added
    automatically to every page.

``max_thumbnail_size``
    The file size in bytes at which images will not be used for their icon
    previews on folder browsing pages. Default: 20000.

``www_root``
    URL path prefix to the root of DXR's web app. Example: ``/smoo``. Default:
    empty.


Tree Sections
-------------

Any section not named ``[DXR]`` represents a tree to be indexed. Changes to
per-tree options take effect when the tree is next indexed.

``build_command``
    Command for building your source code. Default: ``make -j {workers}``.
    This is run within ``object_folder``. Note that ``{workers}`` will be
    replaced with ``workers`` from the ``[DXR]`` section (though 1 if
    ``workers`` is set to 0).

``clean_command``
    Command for deleting the build products of ``build_command``, restoring
    things to the pre-built state. Default: ``make clean``. This is run within
    ``object_folder``.

``disabled_plugins``
   Plugins disabled in this tree, in addition to ones already disabled in the
   ``[DXR]`` section. Default: ``*``

``enabled_plugins``
    Plugins enabled in this tree. Default: ``*``, which enables the same
    plugins enabled in the ``[DXR]`` section.

``es_shards``
    The number of shards to break the elasticsearch index into. Default: 5

``ignore_patterns``
    Whitespace-separated list of Unix `shell-style
    <http://docs.python.org/library/fnmatch.html>`__ file names or paths to
    ignore. Paths start with a slash, and file names don't. Patterns
    containing whitespace can be expressed by enclosing them in double quotes:
    ``"Lovely readable name.human"``.

``object_folder``
    Folder where the ``build_command`` will be run. This is generally the
    folder where object files will be stored. Default: same as
    ``source_folder``

``source_folder``
    The folder containing the source code to index. **Required.**

``source_encoding``
    The Unicode encoding of the tree's source files. Default: ``utf-8``

``temp_folder``
    A ``format()``-style template for deciding where to store temporary files
    used while indexing. The token ``{tree}`` will be replaced with the name
    of each tree you index. Default: ``temp_folder`` setting from ``[DXR]``
    section. You generally don't need to set this.

``p4web_url``
    The URL to the root of a p4web installation. Default: ``http://p4web/``

``workers``
    Number of concurrent processes to use for building and indexing this tree.
    Default: ``workers`` setting from ``[DXR]`` section. You might want to set
    this lower for a tree that uses memory-hungry plugins if you're low on RAM.

Plugin Configuration
====================

Plugin-specific options go in ``[[double-bracketed]]`` sections under trees.
For example... ::

    [some-tree]

        [[buglink]]
        url = http://www.example.com/
        name = Example bug tracker

Currently, changes to plugin configuration take effect at index time or after
restarting the web app; none are picked up by the web app in realtime.

See :ref:`writing-plugins` for more details on plugin development.

[[buglink]]
-----------

``name``
    Name of the tree's bug tracker installation, e.g. ``Mozilla's Bugzilla``

``regex``
    Regex for finding bug references to link in the source code. Default:
    ``(?i)bug\s+#?([0-9]+)``, which catches things like "bug 123456"

``url``
    URL pattern for building links to tickets. ``%s`` will be replaced with the
    ticket number. The option should include the URL scheme.

[[python]]
----------

``python_path``
    Path to the folder from which the codebase imports Python modules

[[xpidl]]
---------

``header_path``
    Path to the folder where generated .h headers will be placed, used for
    URL construction.

``include_folders``
    Whitespace-separated list of paths to search in to resolve include
    directives. Default: [] (current folder)


Syntax
=======

``#`` comments out the remainder of its line in most cases. To express a config
value that contains ``#``, place it in triple quotes::

    regex = '''(?i)bug\s+\#?([0-9]+)'''
    regex = """(?i)bug\s+\#?([0-9]+)"""

A single surrounding pair of single or double quotes will end up as part of the
value at the moment, due to an apparent `bug in configobj
<https://github.com/DiffSK/configobj/issues/97>`_.


docs/source/example-configuration.rst
::

    [DXR]
    # Some global options here, if you like

    [yourproject]
    source_folder       = /code/my-checkout
    build_command       = make clean; make -j {workers}


dxr/config.py
"""Configuration file abstractions

Please update docs/source/configuration.rst when you change this.

"""
from datetime import datetime
from multiprocessing import cpu_count
from ordereddict import OrderedDict
from os import getcwd
from os.path import abspath

from configobj import ConfigObj
from funcy import merge
from more_itertools import first
from pkg_resources import resource_string
from schema import Optional, Use, And, Schema, SchemaError

from dxr.exceptions import ConfigError
from dxr.plugins import all_plugins_but_core, core_plugin
from dxr.utils import cd, if_raises


# Format version, signifying the instance format this web frontend code is
# able to serve. Must match exactly; deploy will do nothing until it does.
FORMAT = resource_string('dxr', 'format').strip()

WORKERS_VALIDATOR = And(Use(int),
                        lambda v: v >= 0,
                        error='"workers" must be a non-negative integer.')


class DotSection(object):
    """In the absense of an actual attribute, let attr lookup fall through to
    ``self._section[attr]``."""

    def __getattr__(self, attr):
        if not hasattr(self, '_section'):  # Happens during unpickling
            raise AttributeError(attr)
        try:
            val = self._section[attr]
        except KeyError:
            raise AttributeError(attr)
        if isinstance(val, dict):
            # So we can dot into nested dicts
            return DotSectionWrapper(val)
        return val


class DotSectionWrapper(DotSection):
    """A wrapper for non-first-level sections so we can dot our way through
    them as well"""

    def __init__(self, section):
        self._section = section


class Config(DotSection):
    """Validation and encapsulation for the DXR config file

    Examples::

        # Settings from the [DXR] section:
        >>> Config(...).default_tree

        # Settings from individual trees:
        >>> Config(...).trees['some-tree'].build_command

        # Settings from plugin-specific sections of trees:
        >>> Config(...).trees['some-tree'].buglink.url

    """
    # Design decisions:
    # * Keep the explicit [DXR] section name because, otherwise, global options
    #   can collide with tree names.
    # * Keep tree.config shortcut. It makes a lot of things shorter and doesn't
    #   hurt anything, since it's all read-only anyway.
    # * Crosswire __getattr__ with __getitem__. It makes callers more readable.
    # * Keep whitespace-delimited lists in config file. I prefer them to
    #   commas, people will have to change less, and it wasn't a big deal to
    #   customize.
    # * Use configobj to parse the ini file and then schema to validate it.
    #   configobj's favored "validate" module makes it very hard to have
    #   whitespace-delimited lists and doesn't compose well since its
    #   validators are strings. The downside is we only report 1 error at a
    #   time, but this can and should be fixed in the schema lib.

    def __init__(self, input, relative_to=None):
        """Pull in and validate a config file.

        :arg input: A string or dict from which to populate the config
        :arg relative_to: The dir relative to which to interpret relative paths

        Raise ConfigError if the configuration is invalid.

        """
        schema = Schema({
            'DXR': {
                Optional('temp_folder', default=abspath('dxr-temp-{tree}')):
                    AbsPath,
                Optional('default_tree', default=None): basestring,
                Optional('disabled_plugins', default=plugin_list('')): Plugins,
                Optional('enabled_plugins', default=plugin_list('*')): Plugins,
                Optional('generated_date',
                         default=datetime.utcnow()
                                         .strftime("%a, %d %b %Y %H:%M:%S +0000")):
                    basestring,
                Optional('log_folder', default=abspath('dxr-logs-{tree}')):
                    AbsPath,
                Optional('workers', default=if_raises(NotImplementedError,
                                                      cpu_count,
                                                      1)):
                    WORKERS_VALIDATOR,
                Optional('skip_stages', default=[]): WhitespaceList,
                Optional('www_root', default=''): Use(lambda v: v.rstrip('/')),
                Optional('google_analytics_key', default=''): basestring,
                Optional('es_hosts', default='http://127.0.0.1:9200/'):
                    WhitespaceList,
                # A semi-random name, having the tree name and format version in it.
                Optional('es_index', default='dxr_{format}_{tree}_{unique}'):
                    basestring,
                Optional('es_alias', default='dxr_{format}_{tree}'):
                    basestring,
                Optional('es_catalog_index', default='dxr_catalog'):
                    basestring,
                Optional('es_catalog_replicas', default=1):
                    Use(int, error='"es_catalog_replicas" must be an integer.'),
                Optional('max_thumbnail_size', default=20000):
                    And(Use(int),
                        lambda v: v >= 0,
                        error='"max_thumbnail_size" must be a non-negative '
                              'integer.'),
                Optional('es_indexing_timeout', default=60):
                    And(Use(int),
                        lambda v: v >= 0,
                        error='"es_indexing_timeout" must be a non-negative '
                              'integer.'),
                Optional('es_indexing_retries', default=0):
                    And(Use(int),
                        lambda v: v >= 0,
                        error='"es_indexing_retries" must be a non-negative '
                              'integer.'),
                Optional('es_refresh_interval', default=60):
                    Use(int, error='"es_refresh_interval" must be an integer.')
            },
            basestring: dict
        })

        # Parse the ini into nested dicts:
        config_obj = ConfigObj(input.splitlines() if isinstance(input,
                                                                basestring)
                               else input,
                               list_values=False)

        if not relative_to:
            relative_to = getcwd()
        with cd(relative_to):
            try:
                config = schema.validate(config_obj.dict())
            except SchemaError as exc:
                raise ConfigError(exc.code, ['DXR'])

            self._section = config['DXR']

            # Normalize enabled_plugins:
            if self.enabled_plugins.is_all:
                # Then explicitly enable anything that isn't explicitly
                # disabled:
                self._section['enabled_plugins'] = [
                        p for p in all_plugins_but_core().values()
                        if p not in self.disabled_plugins]

            # Now that enabled_plugins and the other keys that TreeConfig
            # depends on are filled out, make some TreeConfigs:
            self.trees = OrderedDict()  # name -> TreeConfig
            for section in config_obj.sections:
                if section != 'DXR':
                    try:
                        self.trees[section] = TreeConfig(section,
                                                         config[section],
                                                         config_obj[section].sections,
                                                         self)
                    except SchemaError as exc:
                        raise ConfigError(exc.code, [section])

        # Make sure default_tree is defined:
        if not self.default_tree:
            self._section['default_tree'] = first(self.trees.iterkeys())

        # These aren't intended for actual use; they're just to influence
        # enabled_plugins of trees, and now we're done with them:
        del self._section['enabled_plugins']
        del self._section['disabled_plugins']


class TreeConfig(DotSectionWrapper):
    def __init__(self, name, unvalidated_tree, sections, config):
        """Fix up settings that depend on the [DXR] section or have
        inter-setting dependencies. (schema can't do multi-setting validation
        yet, and configobj can't do cross-section interpolation.)

        Add a ``config`` attr to trees as a shortcut back to the [DXR] section
        and a ``name`` attr to save cumbersome tuple unpacks in callers.

        """
        self.config = config
        self.name = name

        schema = Schema({
            Optional('build_command', default='make -j {workers}'): basestring,
            Optional('clean_command', default='make clean'): basestring,
            Optional('description', default=''): basestring,
            Optional('disabled_plugins', default=plugin_list('')): Plugins,
            Optional('enabled_plugins', default=plugin_list('*')): Plugins,
            Optional('es_index', default=config.es_index): basestring,
            Optional('es_shards', default=5):
                Use(int, error='"es_shards" must be an integer.'),
            Optional('ignore_patterns',
                     default=['.hg', '.git', 'CVS', '.svn', '.bzr',
                              '.deps', '.libs', '.DS_Store', '.nfs*', '*~',
                              '._*']): WhitespaceList,
            Optional('object_folder', default=None): AbsPath,
            'source_folder': AbsPath,
            Optional('source_encoding', default='utf-8'): basestring,
            Optional('temp_folder', default=None): AbsPath,
            Optional('p4web_url', default='http://p4web/'): basestring,
            Optional('workers', default=None): WORKERS_VALIDATOR,
            Optional(basestring): dict})
        tree = schema.validate(unvalidated_tree)

        if tree['temp_folder'] is None:
            tree['temp_folder'] = config.temp_folder
        if tree['object_folder'] is None:
            tree['object_folder'] = tree['source_folder']
        if tree['workers'] is None:
            tree['workers'] = config.workers

        # Convert enabled_plugins to a list of plugins:
        if tree['disabled_plugins'].is_all:
            # * doesn't really mean "all" in a tree. It means "everything the
            # [DXR] section enabled".
            tree['disabled_plugins'] = config.enabled_plugins
        else:
            # Add anything globally disabled to our local disabled list:
            tree['disabled_plugins'].extend(p for p in config.disabled_plugins
                                            if p not in
                                            tree['disabled_plugins'])

        if tree['enabled_plugins'].is_all:
            tree['enabled_plugins'] = [p for p in config.enabled_plugins
                                       if p not in tree['disabled_plugins']]
        tree['enabled_plugins'].insert(0, core_plugin())

        # Split ignores into paths and filenames:
        tree['ignore_paths'] = [i for i in tree['ignore_patterns']
                                if i.startswith('/')]
        tree['ignore_filenames'] = [i for i in tree['ignore_patterns']
                                    if not i.startswith('/')]

        # Delete misleading, useless, or raw values people shouldn't use:
        del tree['ignore_patterns']
        del tree['disabled_plugins']

        # Validate plugin config:
        enableds_with_all_optional_config = set(
            p for p in tree['enabled_plugins']
            if all(isinstance(k, Optional) for k in p.config_schema.iterkeys()))
        plugin_schema = Schema(merge(
            dict((Optional(name) if plugin in enableds_with_all_optional_config
                                    or plugin not in tree['enabled_plugins']
                                 else name,
                  plugin.config_schema)
                 for name, plugin in all_plugins_but_core().iteritems()),
            # And whatever isn't a plugin section, that we don't care about:
            {object: object}))
        # Insert empty missing sections for enabled plugins with entirely
        # optional config so their defaults get filled in. (Don't insert them
        # if the plugin has any required options; then we wouldn't produce the
        # proper error message about the section being absent.)
        for plugin in enableds_with_all_optional_config:
            tree.setdefault(plugin.name, {})
        tree = plugin_schema.validate(tree)

        super(TreeConfig, self).__init__(tree)

    @property
    def log_folder(self):
        """Return the global log_folder with the tree name subbed in."""
        return self.config.log_folder.format(tree=self.name)

    @property
    def temp_folder(self):
        """Return ``self.temp_folder`` with the tree name subbed in."""
        return self.config.temp_folder.format(tree=self.name)


class ListAndAll(list):
    """A list we can also store an ``all`` attr on, indicating whether it
    was derived from a configuration value of ``*``"""


def plugin_list(value):
    """Turn a space-delimited series of plugin names into a ListAndAll of
    Plugins.

    Never return the core plugin.

    """
    if not isinstance(value, basestring):
        # Probably can't happen
        raise SchemaError('"%s" is neither * nor a whitespace-delimited list '
                          'of plugin names.' % (value,), None)

    plugins = all_plugins_but_core()
    names = value.strip().split()
    is_all = names == ['*']
    if is_all:
        names = plugins.keys()
    try:
        ret = ListAndAll([plugins[name] for name in names])
        ret.is_all = is_all
        return ret
    except KeyError:
        raise SchemaError('Never heard of plugin "%s". I\'ve heard of '
                          'these: %s.' % (name, ', '.join(plugins.keys())),
                          None)
Plugins = Use(plugin_list)


WhitespaceList = And(basestring,
                     Use(lambda value: value.strip().split()),
                     error='This should be a whitespace-separated list.')


# Turn a filesystem path into an absolute one so changing the working
# directory doesn't keep us from finding them.
AbsPath = And(basestring, Use(abspath), error='This should be a path.')
